# When Additive and Multiplicative Inverses are the Same

Names: Carlos, Sahib, Quang



---



# Introduction
In our abstract algebra and number theory classes, we delve into the process of modulo arithmetic, specifically focusing on multiplicative inverses and additive inverses.

Inspiration for the paper on this idea:

An intriguing incident occurred at a university in Connecticut. A student, while calculating the multiplicative and additive inverses for a ring, discovered that the multiplicative inverse of 43 modulo 50 is 7. Interestingly, the professor had initially thought that the answer should be the additive inverse. However, it turned out that both the additive and multiplicative inverses of 43 modulo 50 were indeed 7.[1]

This unexpected coincidence served as the motivation our thier final project paper, where we further explore this phenomenon.

## Prerequiste for existance of inverse pairs

First we need to load library and create a function for create libraries and data frame for better looking table python resources

In [ ]:
import math   # for sqrt and other math functions
import time   # for runtime analysis
import matplotlib.pyplot as plt  # for plotting
import pandas as pd
from sympy import factorint
import tabulate as table
from IPython.display import display_html
from itertools import chain,cycle
def display_side_by_side(*args,titles=cycle([''])):
    html_str=''
    for df,title in zip(args, chain(titles,cycle(['</br>'])) ):
        html_str+='<th style="text-align:center"><td style="vertical-align:top">'
        html_str+=f'<h2 style="text-align: center;">{title}</h2>'
        html_str+=df.to_html().replace('table','table style="display:inline"')
        html_str+='</td></th>'
    display_html(html_str,raw=True)#Code cited from https://stackoverflow.com/questions/38783027/jupyter-notebook-display-two-pandas-tables-side-by-side



We will begin by providing computation for definition 1 in the paper. Definition 1 reads as such:

**Definition 1.** A Pythagorean prime is a prime of the form $4n + 1$ for $n \in \mathbb{Z^+}$, or an odd prime that is expressible as the sum of two squares.

First we need a isPrime code that checks if prime and returns true or false.

In [ ]:
def isPrime(n) :
  for m in range(2,n): #specifies range of integers to check divisibility of our chosen prime n
   #print(m)
   if  n%m==0: #chceks the divisibility, with divisibile then not prime
      #print("Not Prime")
      return False
  else: #otherwise it is prime
    #print("Prime")
    return True
####
isPrime(10)
isPrime(13)

True

Additionally, we need a module that checks the pythogorean prime condition. This will rely on the two conditions as the definition states.

In [ ]:
def pythoprime (n):
  if(isPrime(n) and (n % 4 == 1)): #checks the condition for pythogorean primes (when prime and n mod 4 = 1 then it satisfies criteria)
    print("YES")
  else:
    print("NO")

In [ ]:
pythoprime(5)

YES


In [ ]:
pythoprime(7)

NO


Although the pythoprime does tell us wether a number $n$ is a pythogorean prime, it would be more beneficial if our module created a list of pythogorean primes. This is what we will do now.

In [ ]:
def pythoprimelist(n):
    pythoprimetable = [] #empty pythogorean prime list to hold values
    if isPrime(n) and (n % 4 == 1): #checks codition again
        pythoprimetable.append(n) #if true adds to the list
    else:
        pythoprimetable.append(0) #otherwise appends 0 to list
    return pythoprimetable # Always return a list OR Check if the list is not empty??????????
      #print(pythoprimetable, end=" ")  # Print the list in one line with a space as the separator

In [ ]:
pythoprimetable=[pythoprimelist(n) for n in range(2,5000,1)] #list comprehention that produces the pythogorean prime list
print(pythoprimetable)

[[0], [0], [0], [5], [0], [0], [0], [0], [0], [0], [0], [13], [0], [0], [0], [17], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [29], [0], [0], [0], [0], [0], [0], [0], [37], [0], [0], [0], [41], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [53], [0], [0], [0], [0], [0], [0], [0], [61], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [73], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [89], [0], [0], [0], [0], [0], [0], [0], [97], [0], [0], [0], [101], [0], [0], [0], [0], [0], [0], [0], [109], [0], [0], [0], [113], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [137], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [149], [0], [0], [0], [0], [0], [0], [0], [157], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [173], [0], [0], [0], [0], [0], [0], [0], [181], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [193], [0], [0], [

For the purpose of computing the pythogorean primes, we have accomplished producing a list that will later be used for pythogorean prime exploration. We will now write code that computes definition 2 from the paper.


---



We will need the following two definitions for most of our exploration from here on out.

Definition of set $U(n)$: $U(n) = \{ 0 < x <= n  | gcd(x, n) = 1\}$.

**Defintion 2.** If $a$ is a unit modulo $n$, then we call a a quadratic residue modulo $n$ if there exists an $x$ such that $x^2 \equiv a ( mod \ n)$. If no such $x$ exists, then we call a $a$ quadratic non-residue modulo $n$.

From this defnition, the author notices that if we want to find multiplicative inverses or even better inverse pairs, we need an element such that $a^2+1 ≡ 0 \ (mod \ n).$

Let $a\in \mathbb{Z}_n$ such that $a^{-1} \cdot a \equiv 1 (\ mod\ n)$ must be an element of the multiplicative group of unit, $U(n)$.
It easy to see that $a^{-1} \cdot a \equiv 1 (\ mod\ n)$ is the same as $a^2 \equiv -1 (\ mod\ n)$ or $a^2 +1 ≡ 0 (\ mod\ n)$.

In [ ]:
def find_inverse_pairs(n):
    inverse_pairs = [] #empty list to hold the inverse pairs
    for a in range(1, n):
      if (a*a + 1) % n == 0:  # Check if a is a solution to x^2 + 1 ≡ 0 (mod n)
        if a<n-a: #condition to not append the inverse pairs twice in both asscending and descending order (1st element is always < than senod element of inverse pair)
          inverse_pairs.append((a, n-a))  # Add the pair (a, -a) to the list
    return inverse_pairs #retuns the inverse pairs list

Now we can test the function with examples from the paper to see if we get appropriate inverse pairs.

In [ ]:
# Test the function
n = 10
print(f"Inverse pairs for U({n}): {find_inverse_pairs(n)}")

Inverse pairs for U(10): [(3, 7)]


In [ ]:
# Test the function
n = 53
print(f"Inverse pairs for U({n}): {find_inverse_pairs(n)}")

Inverse pairs for U(53): [(23, 30)]


In [ ]:
n = 32045
print(f"Inverse pairs for U({n}): {find_inverse_pairs(n)}")

Inverse pairs for U(32045): [(2163, 29882), (7093, 24952), (8572, 23473), (9703, 22342), (11003, 21042), (13502, 18543), (14633, 17412), (15933, 16112)]


In [ ]:
n=106
print(f"Inverse pairs for U({n}): {find_inverse_pairs(n)}")

Inverse pairs for U(106): [(23, 83)]


Given our code is now working, we can now explore inverse pairs and cyclic properties and replicate what some of the the paper explored.

# Inverse Pair Exploration

For purposes of not having 0's in our pythogorean prime list, we have modified the code to only append the pythogorean primes to a list, this way we can continue our inverse pair exploration with better efficiency.

In [ ]:
def pythoprimelistcyclic(r): #r specifies the range of integers that we can check for pythogorean primes
    pythoprimetable = [] #empty list that holds pythogorean primes
    scraptable= [] #empty list that will hold none pythogorean primes
    for n in range(2,r):
      if isPrime(n) and (n % 4 == 1):
        pythoprimetable.append(n) #appends the pythogorean primes to the list
      else:
        scraptable.append(0) #appends non pythogorean primes
    return pythoprimetable # Always return a list OR Check if the list is not empty??????????
      #print(pythoprimetable, end=" ")  # Print the list in one line with a space as the separator

In [ ]:
pythoprimelistcyclic(50) #testing

[5, 13, 17, 29, 37, 41]

Additionally, we want to explore the case for $2p$, that is $2\times$ the pythogorean primes. We will create a module that allows you to vary the coeffcient that we can mulitply the pythogorean primes by.

In [ ]:
def multiplepythoprimelist(r,m):
  plist = pythoprimelistcyclic(r) #creates list of the pythogorean primes with range r
  newlist=[] #will hold new values of m * pythogorean primes
  for n in range(len(plist)):
    p =plist[n]*m #computes m * pythogorean primes
    newlist.append(p) #creates the new m*p list
  return newlist #returns the newlist

In [ ]:
multiplepythoprimelist(50,2) #testing

[10, 26, 34, 58, 74, 82]

Now that we have working code do create the $2*p$ list, we can now explore and replicate the modulo/inverse pair tables that represent when the pairs  are both additive and multiplicative inverses.

In [ ]:
m=1 #creates for p inverses
plist=multiplepythoprimelist(200,m)
onepinversepairs=[find_inverse_pairs(plist[n]) for n in range(0,len(plist))]
print(onepinversepairs)

[[(2, 3)], [(5, 8)], [(4, 13)], [(12, 17)], [(6, 31)], [(9, 32)], [(23, 30)], [(11, 50)], [(27, 46)], [(34, 55)], [(22, 75)], [(10, 91)], [(33, 76)], [(15, 98)], [(37, 100)], [(44, 105)], [(28, 129)], [(80, 93)], [(19, 162)], [(81, 112)], [(14, 183)]]


Now we can replicate the tables that are present on page 307 of the paper. That is we can show the additive and mulitplicative invereses of the moduli $p$ and $2*p$.

In [ ]:
m=1
plist=multiplepythoprimelist(200,m)
onepinversepairs=[find_inverse_pairs(plist[n]) for n in range(0,len(plist))]

df = pd.DataFrame({ #prints the dataframe of the pythogorean primes and the inversepairs of 2p
    'modulo': plist, #column one is the mudulo
    'inverse pairs': onepinversepairs #column two is the inverse pairs
})
display_side_by_side(df) #displays the dataframe nicely

modulo 
 inverse pairs 
 
 
 
 
 0 
 5 
 [(2, 3)] 
 
 
 1 
 13 
 [(5, 8)] 
 
 
 2 
 17 
 [(4, 13)] 
 
 
 3 
 29 
 [(12, 17)] 
 
 
 4 
 37 
 [(6, 31)] 
 
 
 5 
 41 
 [(9, 32)] 
 
 
 6 
 53 
 [(23, 30)] 
 
 
 7 
 61 
 [(11, 50)] 
 
 
 8 
 73 
 [(27, 46)] 
 
 
 9 
 89 
 [(34, 55)] 
 
 
 10 
 97 
 [(22, 75)] 
 
 
 11 
 101 
 [(10, 91)] 
 
 
 12 
 109 
 [(33, 76)] 
 
 
 13 
 113 
 [(15, 98)] 
 
 
 14 
 137 
 [(37, 100)] 
 
 
 15 
 149 
 [(44, 105)] 
 
 
 16 
 157 
 [(28, 129)] 
 
 
 17 
 173 
 [(80, 93)] 
 
 
 18 
 181 
 [(19, 162)] 
 
 
 19 
 193 
 [(81, 112)] 
 
 
 20 
 197 
 [(14, 183)]

In [ ]:
m=2
plist=multiplepythoprimelist(200,m)
twopinversepairs=[find_inverse_pairs(plist[n]) for n in range(0,len(plist))]

df = pd.DataFrame({ #prints the dataframe of the pythogorean primes and the inversepairs of 2p
    'modulo': plist, #column one is the mudulo
    'inverse pairs': twopinversepairs #column two is the inverse pairs
})
display_side_by_side(df)

modulo 
 inverse pairs 
 
 
 
 
 0 
 10 
 [(3, 7)] 
 
 
 1 
 26 
 [(5, 21)] 
 
 
 2 
 34 
 [(13, 21)] 
 
 
 3 
 58 
 [(17, 41)] 
 
 
 4 
 74 
 [(31, 43)] 
 
 
 5 
 82 
 [(9, 73)] 
 
 
 6 
 106 
 [(23, 83)] 
 
 
 7 
 122 
 [(11, 111)] 
 
 
 8 
 146 
 [(27, 119)] 
 
 
 9 
 178 
 [(55, 123)] 
 
 
 10 
 194 
 [(75, 119)] 
 
 
 11 
 202 
 [(91, 111)] 
 
 
 12 
 218 
 [(33, 185)] 
 
 
 13 
 226 
 [(15, 211)] 
 
 
 14 
 274 
 [(37, 237)] 
 
 
 15 
 298 
 [(105, 193)] 
 
 
 16 
 314 
 [(129, 185)] 
 
 
 17 
 346 
 [(93, 253)] 
 
 
 18 
 362 
 [(19, 343)] 
 
 
 19 
 386 
 [(81, 305)] 
 
 
 20 
 394 
 [(183, 211)]

As part of exploration, we  decided to try different multiples of the pythogorean primes to see if the same occurance happened where one value of the moduli $p$ appeared in the moduli $m*p$ inverse pairs. Upon investigation we found that the same occured for moduli $5*p$ and $10*p$. The data frames can be found below.

In [ ]:
m=5
plist=multiplepythoprimelist(200,m)
print(plist)
multiplepinversepairs=[find_inverse_pairs(plist[n]) for n in range(0,len(plist))]
print(multiplepinversepairs)

[25, 65, 85, 145, 185, 205, 265, 305, 365, 445, 485, 505, 545, 565, 685, 745, 785, 865, 905, 965, 985]
[[(7, 18)], [(8, 57), (18, 47)], [(13, 72), (38, 47)], [(12, 133), (17, 128)], [(43, 142), (68, 117)], [(32, 173), (73, 132)], [(23, 242), (83, 182)], [(72, 233), (133, 172)], [(27, 338), (173, 192)], [(123, 322), (212, 233)], [(22, 463), (172, 313)], [(192, 313), (212, 293)], [(33, 512), (142, 403)], [(98, 467), (128, 437)], [(37, 648), (237, 448)], [(193, 552), (342, 403)], [(28, 757), (342, 443)], [(93, 772), (253, 612)], [(162, 743), (343, 562)], [(112, 853), (467, 498)], [(183, 802), (408, 577)]]


In [ ]:
# Create a DataFrame
df = pd.DataFrame({
    'modulo': plist,
    'inverse pairs': multiplepinversepairs
})
display_side_by_side(df)

modulo 
 inverse pairs 
 
 
 
 
 0 
 25 
 [(7, 18)] 
 
 
 1 
 65 
 [(8, 57), (18, 47)] 
 
 
 2 
 85 
 [(13, 72), (38, 47)] 
 
 
 3 
 145 
 [(12, 133), (17, 128)] 
 
 
 4 
 185 
 [(43, 142), (68, 117)] 
 
 
 5 
 205 
 [(32, 173), (73, 132)] 
 
 
 6 
 265 
 [(23, 242), (83, 182)] 
 
 
 7 
 305 
 [(72, 233), (133, 172)] 
 
 
 8 
 365 
 [(27, 338), (173, 192)] 
 
 
 9 
 445 
 [(123, 322), (212, 233)] 
 
 
 10 
 485 
 [(22, 463), (172, 313)] 
 
 
 11 
 505 
 [(192, 313), (212, 293)] 
 
 
 12 
 545 
 [(33, 512), (142, 403)] 
 
 
 13 
 565 
 [(98, 467), (128, 437)] 
 
 
 14 
 685 
 [(37, 648), (237, 448)] 
 
 
 15 
 745 
 [(193, 552), (342, 403)] 
 
 
 16 
 785 
 [(28, 757), (342, 443)] 
 
 
 17 
 865 
 [(93, 772), (253, 612)] 
 
 
 18 
 905 
 [(162, 743), (343, 562)] 
 
 
 19 
 965 
 [(112, 853), (467, 498)] 
 
 
 20 
 985 
 [(183, 802), (408, 577)]

In [ ]:
m=10
plist=multiplepythoprimelist(200,m)
print(plist)
multiplep10inversepairs=[find_inverse_pairs(plist[n]) for n in range(0,len(plist))]
print(multiplep10inversepairs)
# Create a DataFrame
df = pd.DataFrame({
    'modulo': plist,
    'inverse pairs': multiplep10inversepairs
})
display_side_by_side(df)

[50, 130, 170, 290, 370, 410, 530, 610, 730, 890, 970, 1010, 1090, 1130, 1370, 1490, 1570, 1730, 1810, 1930, 1970]
[[(7, 43)], [(47, 83), (57, 73)], [(13, 157), (47, 123)], [(17, 273), (133, 157)], [(43, 327), (117, 253)], [(73, 337), (173, 237)], [(23, 507), (83, 447)], [(133, 477), (233, 377)], [(27, 703), (173, 557)], [(123, 767), (233, 657)], [(313, 657), (463, 507)], [(293, 717), (313, 697)], [(33, 1057), (403, 687)], [(437, 693), (467, 663)], [(37, 1333), (237, 1133)], [(193, 1297), (403, 1087)], [(443, 1127), (757, 813)], [(93, 1637), (253, 1477)], [(343, 1467), (743, 1067)], [(467, 1463), (853, 1077)], [(183, 1787), (577, 1393)]]


modulo 
 inverse pairs 
 
 
 
 
 0 
 50 
 [(7, 43)] 
 
 
 1 
 130 
 [(47, 83), (57, 73)] 
 
 
 2 
 170 
 [(13, 157), (47, 123)] 
 
 
 3 
 290 
 [(17, 273), (133, 157)] 
 
 
 4 
 370 
 [(43, 327), (117, 253)] 
 
 
 5 
 410 
 [(73, 337), (173, 237)] 
 
 
 6 
 530 
 [(23, 507), (83, 447)] 
 
 
 7 
 610 
 [(133, 477), (233, 377)] 
 
 
 8 
 730 
 [(27, 703), (173, 557)] 
 
 
 9 
 890 
 [(123, 767), (233, 657)] 
 
 
 10 
 970 
 [(313, 657), (463, 507)] 
 
 
 11 
 1010 
 [(293, 717), (313, 697)] 
 
 
 12 
 1090 
 [(33, 1057), (403, 687)] 
 
 
 13 
 1130 
 [(437, 693), (467, 663)] 
 
 
 14 
 1370 
 [(37, 1333), (237, 1133)] 
 
 
 15 
 1490 
 [(193, 1297), (403, 1087)] 
 
 
 16 
 1570 
 [(443, 1127), (757, 813)] 
 
 
 17 
 1730 
 [(93, 1637), (253, 1477)] 
 
 
 18 
 1810 
 [(343, 1467), (743, 1067)] 
 
 
 19 
 1930 
 [(467, 1463), (853, 1077)] 
 
 
 20 
 1970 
 [(183, 1787), (577, 1393)]

**Conjecture:** We were able to find an element that was a part of an inverse pair for both modulo $5p$ and modulo $10p$, similarly to the paper doing so for modulo $p$ and modulo $2p$.

# Using computation to verify theorem 2 in the paper

As an essential element to the paper, we wanted to provide code that would prove the findings of theorem 2.

**Theorem 2.** If $n = p_1^{r_1}p_2^{r_2}\dots p_k^{r_k}$ or $n = 2p_1^{r_1}p_2^{r_2}\dots p_k^{r_k}$, where for each $1 \leq i \leq k, p_i$ is a Pyhtogorean prime and $r_i$ is apositive integer, then there exist exactly $2^{k-1}$ inverse pairs.

As part of this we needed a primefactors code as a prerequisite to a theorem 2 code that could check if theorem 2 was proven or not. Below is a primeFactors code that we were able to find that provided exactly what we needed to prove theorem 2.

In [ ]:
def primeFactors(n):
	factors = []

	while n % 2 == 0:
		factors.append(2)
		n = n // 2

	# n must be odd at this point
	# so a skip of 2 ( i = i + 2) can be used
	for i in range(3,int(math.sqrt(n))+1,2):

		# while i divides n , print i ad divide n
		while n % i== 0:
			factors.append(i)
			n = n // i

	# Condition if n is a prime
	# number greater than 2
	if n > 2:
		factors.append(n)
	return factors

# This code is contributed by Harshit Agrawal to Geeksforgeeks
#Code Improved by Sarthak Shrivastava from Geeksforgeeks

Now we can test to see if the prime factors code is working and producing the same results as did Example 2, 3, and 4 did on page 302 of the paper.

In [ ]:
n = 425 #example 2
print(primeFactors(n))

[5, 5, 17]


In [ ]:
n = 2210 #example 3
print(primeFactors(n))

[2, 5, 13, 17]


In [ ]:
n = 32045 #example 4
print(primeFactors(n))

[5, 13, 17, 29]


Given the prime factors does exactly what we need to assist in writing code that checks if theorem 2 is upheld, we can now move onto writing this code.

In [ ]:
testlist=pythoprimelistcyclic(5000) #a large list of pythogorean primes that will be used in theorem 2
print(testlist)

[5, 13, 17, 29, 37, 41, 53, 61, 73, 89, 97, 101, 109, 113, 137, 149, 157, 173, 181, 193, 197, 229, 233, 241, 257, 269, 277, 281, 293, 313, 317, 337, 349, 353, 373, 389, 397, 401, 409, 421, 433, 449, 457, 461, 509, 521, 541, 557, 569, 577, 593, 601, 613, 617, 641, 653, 661, 673, 677, 701, 709, 733, 757, 761, 769, 773, 797, 809, 821, 829, 853, 857, 877, 881, 929, 937, 941, 953, 977, 997, 1009, 1013, 1021, 1033, 1049, 1061, 1069, 1093, 1097, 1109, 1117, 1129, 1153, 1181, 1193, 1201, 1213, 1217, 1229, 1237, 1249, 1277, 1289, 1297, 1301, 1321, 1361, 1373, 1381, 1409, 1429, 1433, 1453, 1481, 1489, 1493, 1549, 1553, 1597, 1601, 1609, 1613, 1621, 1637, 1657, 1669, 1693, 1697, 1709, 1721, 1733, 1741, 1753, 1777, 1789, 1801, 1861, 1873, 1877, 1889, 1901, 1913, 1933, 1949, 1973, 1993, 1997, 2017, 2029, 2053, 2069, 2081, 2089, 2113, 2129, 2137, 2141, 2153, 2161, 2213, 2221, 2237, 2269, 2273, 2281, 2293, 2297, 2309, 2333, 2341, 2357, 2377, 2381, 2389, 2393, 2417, 2437, 2441, 2473, 2477, 2521, 2549,

In [ ]:
def theorem2(n):
    primefactor = list(set(primeFactors(n))) #list of the primefactors
    print(primefactor)
    pythoprimetable = testlist #calls upon our large pythogorean prime list
    inverse_pairs = find_inverse_pairs(n)  # Use your function to find inverse pairs

    pythoprime_count = 0 #will count the number of pythogorean primes that show up in our factorization list
    for p in primefactor:
        if p in pythoprimetable:
            pythoprime_count += 1

    k = pythoprime_count  # k is now the count of unique Pythagorean primes
    print(k)
    if len(inverse_pairs) == 2**(k-1):  # Compare with the number of inverse pairs
        return True
    else:
        return False

In [ ]:
print(theorem2(425)) #example 1

[17, 5]
2
True


In [ ]:
print(theorem2(2210)) #example 2

[17, 2, 13, 5]
3
True


In [ ]:
print(theorem2(32045)) #example 3

[17, 13, 29, 5]
4
True


In [ ]:
print(theorem2(11)) #test for non pythogorean primes

[11]
0
False


We were able to provide code that verifies that theorem 2 is satisfied by using the examples from the paper. Additionally, it does not work if using other numbers that aren't pythogorean which in return do not satisfy the theorem.

# Finding Cyclic Group

As part of the paper and using DAVID R. GUICHARD paper "When Is $U(n)$ Cyclic?", we decided to create code that found when something is cyclic and to return the genreator of the cyclic group for $U(n)$.

**Primitive Root Theorem:** $U(n)$ is cyclic if and only if $n$ is $1, 2, 4, p^k,$ or $2p^k$, where p  is an odd prime and $k \geq 1$.

We want to prove this theorem so that we enable to show that $∀ n$ that is a pythagorean prime, then $U(n)$ will be cylic.

We will need code that gives us the generators for cyclic $U$ groups. We want to make a function to show the Primitive Root Theorem. First, we must create our $U$ groups.

In [ ]:
def U(n):
    return [i for i in range(1, n) if math.gcd(i, n) == 1]

In [ ]:
U(7)

[1, 2, 3, 4, 5, 6]

Now we can write a check cyclic code that will find the generator of our $U(mod)$.

In [ ]:
def realcyclicalgorithm(mod):
  ulist= U(mod) #creates the list for the U group
  checklist=[] #holds the checks
  for n in range(len(ulist)):
    for j in range(1,len(ulist)+1):
      element=ulist[n]
      #print("element", element)
      powered=ulist[n]**j #takes element of U(n) and raises to power j
      #print("power",j)
      #print("powered", powered)
      check = powered % mod #takes mod of the ulist[n]^j
      #print("check",check)
      checklist.append(check) #holds the moded value in new list
      #print("checklist",checklist)
      if check == 1 and j != len(ulist)-1: #if 1 breaks because not cyclic yet and when j!= max value
        break
    if len(ulist)==len(checklist): #if len of checklist is equivilant to ulist then we have element that is our generator
      if ulist[n]==1:
        print(f"U({mod}) is not cyclic")
        break
      print(f"{ulist[n]} is the generator of U({mod})")
      break
    else:
      checklist.clear()

In [ ]:
realcyclicalgorithm(7)

3 is the generator of U(7)


In [ ]:
realcyclicalgorithm(13)

2 is the generator of U(13)


Now we can test for a bunch of pythagorean prime numbers and see what their generators are, and whether all are cyclic.

## Pythogorean Prime Cyclic Groups

In [ ]:
def pythoprimelistcyclic(r):
    pythoprimetable = []
    scraptable= []
    for n in range(r):
      if isPrime(n) and (n % 4 == 1):
        pythoprimetable.append(n)
      else:
        scraptable.append(0)
    return pythoprimetable # Always return a list OR Check if the list is not empty??????????
      #print(pythoprimetable, end=" ")  # Print the list in one line with a space as the separator

In [ ]:
testlist=pythoprimelistcyclic(500)

In [ ]:
def pythoprimecylic(table):
  for n in range(len(table)):
    pythoprime=table[n]
    realcyclicalgorithm(pythoprime)

In [ ]:
pythoprimecylic(testlist)

2 is the generator of U(5)
2 is the generator of U(13)
3 is the generator of U(17)
2 is the generator of U(29)
2 is the generator of U(37)
6 is the generator of U(41)
2 is the generator of U(53)
2 is the generator of U(61)
5 is the generator of U(73)
3 is the generator of U(89)
5 is the generator of U(97)
2 is the generator of U(101)
6 is the generator of U(109)
3 is the generator of U(113)
3 is the generator of U(137)
2 is the generator of U(149)
5 is the generator of U(157)
2 is the generator of U(173)
2 is the generator of U(181)
5 is the generator of U(193)
2 is the generator of U(197)
6 is the generator of U(229)
3 is the generator of U(233)
7 is the generator of U(241)
3 is the generator of U(257)
2 is the generator of U(269)
5 is the generator of U(277)
3 is the generator of U(281)
2 is the generator of U(293)
10 is the generator of U(313)
2 is the generator of U(317)
10 is the generator of U(337)
2 is the generator of U(349)
3 is the generator of U(353)
2 is the generator of U(

As it turns out, all $U(p)$ are cyclic. Now we are curious about doing this for other sequences.

# Modulo of Pentagonal Numbers

First we will create a list of pentagonal numbers.

In [ ]:
def pentagonalnumberlist(r):
  penlist=[]
  for n in range(r):
    p = int((3*n*n-n)/2)
    penlist.append(p)
  return penlist

In [ ]:
pentagonalnumberlist(25)

[0,
 1,
 5,
 12,
 22,
 35,
 51,
 70,
 92,
 117,
 145,
 176,
 210,
 247,
 287,
 330,
 376,
 425,
 477,
 532,
 590,
 651,
 715,
 782,
 852]

In [ ]:
n=200
pentagonaltable=pentagonalnumberlist(n)
pentpinversepairs=[find_inverse_pairs(pentagonaltable[n]) for n in range(0,len(pentagonaltable))]

In [ ]:
df1 = pd.DataFrame({
    'modulo': pentagonaltable,
    'inverse pairs': pentpinversepairs
})
df_filtered1 = df1[df1['inverse pairs'].astype(bool)]
display_side_by_side(df_filtered1)

modulo 
 inverse pairs 
 
 
 
 
 2 
 5 
 [(2, 3)] 
 
 
 10 
 145 
 [(12, 133), (17, 128)] 
 
 
 17 
 425 
 [(132, 293), (157, 268)] 
 
 
 25 
 925 
 [(43, 882), (68, 857)] 
 
 
 34 
 1717 
 [(293, 1424), (616, 1101)] 
 
 
 41 
 2501 
 [(50, 2451), (255, 2246)] 
 
 
 50 
 3725 
 [(193, 3532), (1832, 1893)] 
 
 
 58 
 5017 
 [(945, 4072), (1810, 3207)] 
 
 
 65 
 6305 
 [(463, 5842), (798, 5507), (1477, 4828), (2738, 3567)] 
 
 
 73 
 7957 
 [(1341, 6616), (2801, 5156)] 
 
 
 74 
 8177 
 [(931, 7246), (2189, 5988), (3336, 4841), (3583, 4594)] 
 
 
 97 
 14065 
 [(1433, 12632), (2738, 11327), (2888, 11177), (4193, 9872)] 
 
 
 106 
 16801 
 [(4641, 12160), (6860, 9941)] 
 
 
 113 
 19097 
 [(437, 18660), (9394, 9703)] 
 
 
 122 
 22265 
 [(538, 21727), (5867, 16398), (7492, 14773), (8368, 13897)] 
 
 
 130 
 25285 
 [(1282, 24003), (2608, 22677), (8832, 16453), (12563, 12722)] 
 
 
 137 
 28085 
 [(237, 27848), (2018, 26067), (10997, 17088), (13252, 14833)] 
 
 
 170 
 43265 
 [(208, 43057), (4373, 38892), (12933, 30332), (17098, 26167)] 
 
 
 178 
 47437 
 [(9667, 37770), (11981, 35456), (20860, 26577), (23174, 24263)] 
 
 
 185 
 51245 
 [(4372, 46873), (5877, 45368), (9478, 41767), (19727, 31518)] 
 
 
 193 
 55777 
 [(13043, 42734), (24816, 30961)]

Lets find the inverse pair of module pentagonal numbers 2 times pentagonal number :)

## Modulo 2 $\times$ Pentagonal number

In [ ]:
def twopentagonalnumberlist(r):
  penlist=[]
  for n in range(r):
    p = 2*int((3*n*n-n)/2)
    penlist.append(p)
  return penlist

In [ ]:
twopentagonalnumberlist(25)

[0,
 2,
 10,
 24,
 44,
 70,
 102,
 140,
 184,
 234,
 290,
 352,
 420,
 494,
 574,
 660,
 752,
 850,
 954,
 1064,
 1180,
 1302,
 1430,
 1564,
 1704]

In [ ]:
n=200
twopentagonaltable=twopentagonalnumberlist(n)
twopentpinversepairs=[find_inverse_pairs(twopentagonaltable[n]) for n in range(0,len(twopentagonaltable))]

In [ ]:
df2 = pd.DataFrame({
    'modulo': twopentagonaltable,
    'inverse pairs': twopentpinversepairs
})
df_filtered2 = df2[df2['inverse pairs'].astype(bool)]
display_side_by_side(df_filtered2)

modulo 
 inverse pairs 
 
 
 
 
 2 
 10 
 [(3, 7)] 
 
 
 10 
 290 
 [(17, 273), (133, 157)] 
 
 
 17 
 850 
 [(157, 693), (293, 557)] 
 
 
 25 
 1850 
 [(43, 1807), (857, 993)] 
 
 
 34 
 3434 
 [(293, 3141), (1101, 2333)] 
 
 
 41 
 5002 
 [(255, 4747), (2451, 2551)] 
 
 
 50 
 7450 
 [(193, 7257), (1893, 5557)] 
 
 
 58 
 10034 
 [(945, 9089), (3207, 6827)] 
 
 
 65 
 12610 
 [(463, 12147), (1477, 11133), (3567, 9043), (5507, 7103)] 
 
 
 73 
 15914 
 [(1341, 14573), (2801, 13113)] 
 
 
 74 
 16354 
 [(931, 15423), (2189, 14165), (3583, 12771), (4841, 11513)] 
 
 
 97 
 28130 
 [(1433, 26697), (4193, 23937), (11177, 16953), (11327, 16803)] 
 
 
 106 
 33602 
 [(4641, 28961), (9941, 23661)] 
 
 
 113 
 38194 
 [(437, 37757), (9703, 28491)] 
 
 
 122 
 44530 
 [(5867, 38663), (13897, 30633), (14773, 29757), (21727, 22803)] 
 
 
 130 
 50570 
 [(12563, 38007), (16453, 34117), (22677, 27893), (24003, 26567)] 
 
 
 137 
 56170 
 [(237, 55933), (10997, 45173), (14833, 41337), (26067, 30103)] 
 
 
 170 
 86530 
 [(4373, 82157), (12933, 73597), (26167, 60363), (43057, 43473)] 
 
 
 178 
 94874 
 [(9667, 85207), (11981, 82893), (24263, 70611), (26577, 68297)] 
 
 
 185 
 102490 
 [(5877, 96613), (19727, 82763), (41767, 60723), (46873, 55617)] 
 
 
 193 
 111554 
 [(13043, 98511), (30961, 80593)]

In [ ]:
display_side_by_side(df_filtered1,df_filtered2)

modulo 
 inverse pairs 
 
 
 
 
 2 
 5 
 [(2, 3)] 
 
 
 10 
 145 
 [(12, 133), (17, 128)] 
 
 
 17 
 425 
 [(132, 293), (157, 268)] 
 
 
 25 
 925 
 [(43, 882), (68, 857)] 
 
 
 34 
 1717 
 [(293, 1424), (616, 1101)] 
 
 
 41 
 2501 
 [(50, 2451), (255, 2246)] 
 
 
 50 
 3725 
 [(193, 3532), (1832, 1893)] 
 
 
 58 
 5017 
 [(945, 4072), (1810, 3207)] 
 
 
 65 
 6305 
 [(463, 5842), (798, 5507), (1477, 4828), (2738, 3567)] 
 
 
 73 
 7957 
 [(1341, 6616), (2801, 5156)] 
 
 
 74 
 8177 
 [(931, 7246), (2189, 5988), (3336, 4841), (3583, 4594)] 
 
 
 97 
 14065 
 [(1433, 12632), (2738, 11327), (2888, 11177), (4193, 9872)] 
 
 
 106 
 16801 
 [(4641, 12160), (6860, 9941)] 
 
 
 113 
 19097 
 [(437, 18660), (9394, 9703)] 
 
 
 122 
 22265 
 [(538, 21727), (5867, 16398), (7492, 14773), (8368, 13897)] 
 
 
 130 
 25285 
 [(1282, 24003), (2608, 22677), (8832, 16453), (12563, 12722)] 
 
 
 137 
 28085 
 [(237, 27848), (2018, 26067), (10997, 17088), (13252, 14833)] 
 
 
 170 
 43265 
 [(208, 43057), (4373, 38892), (12933, 30332), (17098, 26167)] 
 
 
 178 
 47437 
 [(9667, 37770), (11981, 35456), (20860, 26577), (23174, 24263)] 
 
 
 185 
 51245 
 [(4372, 46873), (5877, 45368), (9478, 41767), (19727, 31518)] 
 
 
 193 
 55777 
 [(13043, 42734), (24816, 30961)] 
 
 
 
 
 
 
 modulo 
 inverse pairs 
 
 
 
 
 2 
 10 
 [(3, 7)] 
 
 
 10 
 290 
 [(17, 273), (133, 157)] 
 
 
 17 
 850 
 [(157, 693), (293, 557)] 
 
 
 25 
 1850 
 [(43, 1807), (857, 993)] 
 
 
 34 
 3434 
 [(293, 3141), (1101, 2333)] 
 
 
 41 
 5002 
 [(255, 4747), (2451, 2551)] 
 
 
 50 
 7450 
 [(193, 7257), (1893, 5557)] 
 
 
 58 
 10034 
 [(945, 9089), (3207, 6827)] 
 
 
 65 
 12610 
 [(463, 12147), (1477, 11133), (3567, 9043), (5507, 7103)] 
 
 
 73 
 15914 
 [(1341, 14573), (2801, 13113)] 
 
 
 74 
 16354 
 [(931, 15423), (2189, 14165), (3583, 12771), (4841, 11513)] 
 
 
 97 
 28130 
 [(1433, 26697), (4193, 23937), (11177, 16953), (11327, 16803)] 
 
 
 106 
 33602 
 [(4641, 28961), (9941, 23661)] 
 
 
 113 
 38194 
 [(437, 37757), (9703, 28491)] 
 
 
 122 
 44530 
 [(5867, 38663), (13897, 30633), (14773, 29757), (21727, 22803)] 
 
 
 130 
 50570 
 [(12563, 38007), (16453, 34117), (22677, 27893), (24003, 26567)] 
 
 
 137 
 56170 
 [(237, 55933), (10997, 45173), (14833, 41337), (26067, 30103)] 
 
 
 170 
 86530 
 [(4373, 82157), (12933, 73597), (26167, 60363), (43057, 43473)] 
 
 
 178 
 94874 
 [(9667, 85207), (11981, 82893), (24263, 70611), (26577, 68297)] 
 
 
 185 
 102490 
 [(5877, 96613), (19727, 82763), (41767, 60723), (46873, 55617)] 
 
 
 193 
 111554 
 [(13043, 98511), (30961, 80593)]

Interestingly, we find the same relationship with the pentagonal numbers that we did with the inverses of our muduli of pythagorean prime and $2$ times pythagorean prime.

# Second Pentagonal Numbers

In [ ]:
def secondpentagonalnumberlist(r):
  penlist=[]
  for n in range(r):
    p = int((3*n*n+n)/2)
    penlist.append(p)
  return penlist

In [ ]:
secondpentagonalnumberlist(25)

[0,
 2,
 7,
 15,
 26,
 40,
 57,
 77,
 100,
 126,
 155,
 187,
 222,
 260,
 301,
 345,
 392,
 442,
 495,
 551,
 610,
 672,
 737,
 805,
 876]

In [ ]:
n=50
secpentagonaltable=secondpentagonalnumberlist(n)
secpentpinversepairs=[find_inverse_pairs(secpentagonaltable[n]) for n in range(0,len(secpentagonaltable))]

In [ ]:
df = pd.DataFrame({
    'modulo': secpentagonaltable,
    'inverse pairs': secpentpinversepairs
})
print(df)

    modulo             inverse pairs
0        0                        []
1        2                        []
2        7                        []
3       15                        []
4       26                 [(5, 21)]
5       40                        []
6       57                        []
7       77                        []
8      100                        []
9      126                        []
10     155                        []
11     187                        []
12     222                        []
13     260                        []
14     301                        []
15     345                        []
16     392                        []
17     442    [(21, 421), (47, 395)]
18     495                        []
19     551                        []
20     610  [(133, 477), (233, 377)]
21     672                        []
22     737                        []
23     805                        []
24     876                        []
25     950                        []
2

Additionally, we tried it out with Ramsey Numbers and Carmical Numbers.

# Other Sequences

# Ramsey Numbers and Carmichael Numbers

Ramsey Numbers

In [ ]:
ramseylist = [0,1,2, 6, 18, 45, 102, 213, 426, 821, 1538, 2820, 5075, 8996, 15743, 27247, 46709, 79405, 133996, 224640, 374400] #taken from OEIS

In [ ]:
ramseyinversepairs=[find_inverse_pairs(ramseylist[n]) for n in range(0,len(ramseylist))]
print(ramseyinversepairs)

NameError: name 'find_inverse_pairs' is not defined

In [ ]:
dfr = pd.DataFrame({
    'modulo': ramseylist,
    'inverse pairs': ramseyinversepairs
})
df_filteredr = dfr[dfr['inverse pairs'].astype(bool)]
print(df_filteredr)

In [ ]:
pythoprime(821)
pythoprime(1538)
pythoprime(46709)
pythoprime(79405)

**Observation:** It turns out that 4 of the ramsey numbers have inverses that are the same. The first one that has this property is a pythagorean prime where as the others are not.

## Carmichael numbers

Using code that we produced in Math 242: MCM we decided to see if there were any cases for the Carmichael numbers.

In [ ]:
def carmichaelNumbers(n):  #code from math 242
    if n <= 2 or isPrime(n): #positive values and prime
        return False
    factors = factorint(n)
    if any(exp > 1 for exp in factors.values()): #Checing if a number is composite or not
        return False
    for p in factors:
        if (n-1) % (p-1) != 0: #each prime p dividing n, p − 1 also divides n − 1.
            return False
    return True

In [ ]:
carmichaels1 = []
for n in range(3, 100000):
    if carmichaelNumbers(n):
        carmichaels1.append(n)
print(carmichaels1)
print(len(carmichaels1))

In [ ]:
carpairs=[find_inverse_pairs(carmichaels1[n]) for n in range(0,len(carmichaels1))]
#print(carpairs)
# Create a DataFrame
df = pd.DataFrame({
    'modulo': carmichaels1,
    'inverse pairs':carpairs
})
df_filtere = df[df['inverse pairs'].astype(bool)]
print(df_filtere)

In [ ]:
pythoprime(1105)
pythoprime(2465)
pythoprime(10585)
pythoprime(29341)
pythoprime(46657)

**Observation:** It turned out that there are 5 cases for the carmichael numbers in which the additive and mulitplicative inverses are the same. Additionally, these numbers are not pythogorean primes.

# Conclusion

Overall, we were able to replicate a bunch of the ideas that the paper explored. In addition we further explored other sequences and found that there are other numbers that have additive and multiplicative inverese that are the same. Some further exploration could be generalizizing the module of $p$ and $2p$. This would allow us to explore other if there are other theorems, seperate from the primitive root theorem that produce comparable results.

# Works Cited

[1] Briggs, K. S., & Spivey, C. R. (2023). When Additive and Multiplicative Inverses are the Same. Mathematics Magazine, 96(3), 299–307. https://doi.org/10.1080/0025570X.2023.2199700

[2] Guichard, D. R. (1999). When Is U(n) Cyclic? An Algebraic Approach. Mathematics Magazine, 72(2), 139–142. https://doi.org/10.1080/0025570X.1999.11996716